### PR TITLE
Early error for cases in which namedtuple usage causes compile time errors

### PR DESCRIPTION
**Pull request by kumare3**
Signed-off-by: Ketan Umare <ketan.umare@gmail.com>

# TL;DR
The following example is incorrect
```
    nm = typing.NamedTuple("OP", greet=str)

    @task
    def say_hello() -> nm:
        return nm("hello world")

    wf_outputs = typing.NamedTuple("OP2", greet1=str, greet2=str)

    @workflow
    def my_wf() -> wf_outputs:
        # Note only Namedtuples can be created like this
        return wf_outputs(say_hello(), say_hello())
```
This fails at registration time, because Flyte does not support tuples for individual values.

It can be fixed by simply adding `.greet` to the returned value from a task - that returns a namedtuple
```
    @workflow
    def my_wf() -> wf_outputs:
        # Note only Namedtuples can be created like this
        return wf_outputs(say_hello().greet, say_hello().greet)
```
This usually is ok for cases in which you have multiple values returned as the assumption is that you need to de-reference, but for single values, this is not directly obvious. Hence we will cause early failure

## Type
 - [x] Bug Fix - not a bug
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1117
https://github.com/flyteorg/flyte/issues/1139

### CODE REVIEW

File: tests/flytekit/unit/core/test_serialization.py [hunk 1]
```diff
@@ -368,13 +370,51 @@ def parent_wf() -> (int, int, int, int):
 
 def test_serialization_named_outputs_single():
     @task
-    def t1() -> str:
+    def t1() -> typing.NamedTuple("OP", a=str):
         return "Hello"
```
Comment (wild-endeavor) [inline tests/flytekit/unit/core/test_serialization.py:374]: so here we're just wrapping it for them?  not raising an error not forcing them to return `("Hello",)`?

**Review by wild-endeavor — APPROVED** [review]

File: tests/flytekit/unit/core/test_serialization.py (hunk 1 above)
Comment (kumare3) [reply tests/flytekit/unit/core/test_serialization.py:374]: Yup

File: flytekit/core/promise.py [hunk 2]
```diff
@@ -411,6 +418,42 @@ def __str__(self):
...
+    for p in promises:
+        if not isinstance(p, Promise):
+            raise AssertionError(
+                "Workflow outputs can only be promises that are returned by tasks. Found a value of"
+                f"type {type(p)}. Workflows cannot return local variables or constants."
+            )
+        outputs[p.var] = TypeEngine.to_python_value(ctx, p.val, entity_interface.outputs[p.var])
+
+    # Should this class be part of the Interface?
+    t = collections.namedtuple(named_tuple_name, list(outputs.keys()))
```
Comment (katrogan) [inline flytekit/core/promise.py:453]: does ordering of the keys matter here?

Comment (kumare3) [reply flytekit/core/promise.py:453]: No

Comment (kumare3) [reply flytekit/core/promise.py:453]: but it is actually ordered

**Review by katrogan — DISMISSED** [review]

**Review by wild-endeavor — APPROVED** [review]

File: tests/flytekit/unit/core/test_type_hints.py [hunk 3]
```diff
@@ -240,11 +240,13 @@ def my_wf(a: int, b: str) -> (int, str):
...
-    def my_wf3(a: int, b: str) -> int:
-        return (a,)
+    with pytest.raises(AssertionError):
+
+        @workflow
+        def my_wf3(a: int, b: str) -> int:
+            return (a,)
 
-    my_wf3(a=10, b="hello")
+    # my_wf3(a=10, b="hello")
```
Comment (cosmicBboy) [inline tests/flytekit/unit/core/test_type_hints.py:249]: does this need to be uncommented?

Comment (kumare3) [reply tests/flytekit/unit/core/test_type_hints.py:249]: aah should be deleted

**Review by cosmicBboy — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]
